### PR TITLE
Replaced existing `TriangleMesh` class

### PR DESCRIPTION
**Pull request by Andreas-Forster**
The newly introduced `TriangleMesh` class can represent triangulated meshes in 1D, 2D and 3D.

The main changes are:
- The mesh is now a trait parametrized over the dimensions.
- The mesh has now a pointSet as domain and does not extend the Domain[D].
- The mesh has now a triangulation.
- New class BarycentricCoordinates to interpolate in triangles
- New class Interpolator to interpolate between values.
- New class MeshSurfaceProperty
- New class TriangleList
- New file ScalarMeshField where class is extracted from old TriangleMesh file.

The scalismo code is changed to work with the new mesh class. The most important changes are that the access to the domain, and the points is now redirected over the pointSet.

### CODE REVIEW

File: src/main/scala/scalismo/mesh/BarycentricCoordinates.scala [hunk 1]
```diff
@@ -0,0 +1,86 @@
+package scalismo.mesh
```
Comment (marcelluethi) [inline src/main/scala/scalismo/mesh/BarycentricCoordinates.scala]: This file should have a license header

File: src/main/scala/scalismo/mesh/Interpolator.scala [hunk 2]
```diff
@@ -0,0 +1,194 @@
+package scalismo.mesh
```
Comment (marcelluethi) [inline src/main/scala/scalismo/mesh/Interpolator.scala]: This file should have a license header

File: src/main/scala/scalismo/mesh/MeshSurfaceProperty.scala [hunk 3]
```diff
@@ -0,0 +1,157 @@
+package scalismo.mesh
```
Comment (marcelluethi) [inline src/main/scala/scalismo/mesh/MeshSurfaceProperty.scala]: This file should have a license header

File: src/main/scala/scalismo/mesh/ScalarMeshField.scala [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package scalismo.mesh
```
Comment (marcelluethi) [inline src/main/scala/scalismo/mesh/ScalarMeshField.scala]: this file should have a license header

File: src/main/scala/scalismo/mesh/TriangleList.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
+package scalismo.mesh
```
Comment (marcelluethi) [inline src/main/scala/scalismo/mesh/TriangleList.scala]: This file should have a license header

File: src/main/scala/scalismo/mesh/TriangleMesh.scala [hunk 6]
```diff
@@ -15,111 +15,159 @@
...
+}
 
-  // a map that has for every point the neighboring cell ids
-  private[scalismo] val cellMap: mutable.HashMap[PointId, Seq[TriangleCell]] = cellMapOpt.getOrElse(mutable.HashMap())
+object TriangleMesh {
 
-  private[this] def updateCellMapForPtId(ptId: PointId, cell: TriangleCell): Unit = {
-    val cellsForKey = cellMap.getOrElse(ptId, Seq[TriangleCell]())
-    cellMap.update(ptId, cellsForKey :+ cell)
+  def apply[D <: Dim: NDSpace](domain: UnstructuredPointsDomain[D], topology: TriangleList)(implicit creator: Create[D]) = {
```
Comment (marcelluethi) [inline src/main/scala/scalismo/mesh/TriangleMesh.scala]: Shouldn't this be called pointSet?

File: src/main/scala/scalismo/mesh/TriangleMesh.scala [hunk 7]
```diff
@@ -15,111 +15,159 @@
...
-  if (cellMapOpt.isEmpty)
-    for (cell <- cells) {
-      cell.pointIds.foreach(id => updateCellMapForPtId(id, cell))
+  def apply[D <: Dim: NDSpace](points: IndexedSeq[Point[D]], topology: TriangleList)(implicit creator: Create[D]) = {
+    creator.createTriangleMesh(UnstructuredPointsDomain(points.toIndexedSeq), topology)
+  }
+
+  /** Typeclass for creating domains of arbitrary dimensionality */
+  trait Create[D <: Dim] extends CreateUnstructuredPointsDomain[D] {
+    def createTriangleMesh(domain: UnstructuredPointsDomain[D], topology: TriangleList): TriangleMesh[D]
```
Comment (marcelluethi) [inline src/main/scala/scalismo/mesh/TriangleMesh.scala]: As above: PointSet instead of Domain

**Comment by marcelluethi** [comment]
Awesome, thanks. Despite the few minor comments, I would be okay with merging. 
@Ghazi-Bouabene, @thogerig Can you also have a look at it?

**Comment by Andreas-Forster** [comment]
All requested changes implemented.

**Comment by marcelluethi** [comment]
Closed on request of the author, as important information regarding the original author of some of the code was lost. A new PR will be opened.

**Pull request closed without merging**